python/ray/air/execution/resources/resource_manager.py

```diff
@@ -0,0 +1,47 @@
+import abc
+
+from typing import List, Optional
+
+import ray
+from ray.air.execution.resources.request import (
+    ResourceRequest,
+    AllocatedResource,
+)
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class ResourceManager(abc.ABC):
+    def get_resource_futures(self) -> List[ray.ObjectRef]:
+        """Return futures for resources to await."""
+        return []
+
+    def request_resources(self, resource_request: ResourceRequest):
+        """Request resources, e.g. schedule placement group."""
+        raise NotImplementedError
+
+    def cancel_resource_request(self, resource_request: ResourceRequest):
+        """Cancel resource request, e.g. remove placement group."""
+        raise NotImplementedError
+
+    def has_resources_ready(self, resource_request: ResourceRequest) -> bool:
+        """Returns True if resources for the given request are available"""
+        raise NotImplementedError
+
+    def acquire_resources(
```

Comment:
what's the difference between this and `request_resources`?

I first thought that `acquire_resources` is supposed to be the blocking version, but the `Returns None if resources are not available` comment implies that this is actually non-blocking.

Reply:
This follows the explanation in the doc, but I'll add a docstring for this.

```
Because Ray clusters can autoscale, we need to request resources before
    we can allocate them. Effectively this means we have to keep track
    of requested resources (which can trigger autoscaling). Once the requested
    resources are available to be scheduled, the control loop can acquire them.

    This separation also ensures that requests can be cancelled before they resolve
    if resources are not needed anymore.
```


The flow is thus

```
# Create resource manager
        resource_manager = ResourceManager()

        # Create resource request
        resource_request = ResourceRequest([{"CPU": 4}])

        # Pass to resource manager
        resource_manager.request_resources(resource_request)

        # Wait until ready
        while not resource_manager.has_resources_ready(resource_request):
            time.sleep(1)

        # Once ready, acquire resources
        allocated_resources = resource_manager.acquire_resources(resource_request)

        # Bind to remote task or actor
        annotated_remote_fn = allocated_resources.annotate_remote_objects(
            [remote_fn])

        # Run remote function. This will use the acquired resources
        ray.get(annotated_remote_fn.remote())

        # After using the resources, free
        resource_manager.free_resources(annotated_resources)
```

Reply:
Actually is this level of granularity needed today? It seems like the main difference is that you now have a "READY" state between "REQUESTED" and "ACQUIRED", but why not just merge READY/ACQUIRED (and cancel them the same way)?